<!--<div align="center">Семинар № 2-->

<div align="center"><font color=#4682B4 size=4 face=
                          "arial black"><em><strong>Семинар 2</strong></em></font></div>

<div class="alert alert-block alert-info"><em>
    <b>Урок 2. Оценка эффективности инвестиций.</b>
    <br>
    
<b>Задача 1.</b> Есть инвестиционный проект с денежными потоками по кварталам:<br>
Квартал 0 1 2 3 4 5;<br>
денежный поток -1200 100 200 300 400 500;<br>
Необходимо принять решение, инвестируем в проект или нет, если ставка дисконтирования 15% годовых.

Расчёт в таблице ***EXCEL*** :

<img src="ДЗ_22_011.png" width="550" height="200">

По формуле ***EXCEL*** :

<img src="ДЗ_22_012.png" width="550" height="200">

В ***PYTHON***'е :

In [1]:
def net_present_value(discount:list, cashFlow:list, term = False) -> str:
    '''Принимает два списка: discount - список ставок дисконтирования по периодам 
       (например, 10% => 10); cashFlow - список денежный поток по периодам и
       term - третий аргумент(необязательный) любое целое число.
       Выводит срок окупаемости (если указан третий аргумент) в годах, месяцах,
       днях чистую приведенную стоимость '''
     
    if len(discount) != len(cashFlow): return f'Размерности списков не равны. Внимательней!'
    
    disc, npv, da = 1, cashFlow[0], 0
    for _ in range(1, len(cashFlow)):
        disc /= (1 + discount[_] / 100)
        npv0 = npv
        npv += disc * cashFlow[_]
        if term:
            if npv > 0 and (not da):
                da = - npv0 / cashFlow[_] * 365
                print(f'Срок окупаемости : {_ - 1:.0f} года, {da // 30.5:.0f} месяцев,\
 {da % 30.5:.0f} дней;\n'),
               
    if term and npv < 0: print('Проект НЕ окупается за данные периоды\n')
    print(f'Чистая приведенная стоимость равна: {npv:0.2f}') 

In [2]:
discount_rate1 = [15 / 4] * 6
cash_flow1 = [-1200, 100, 200, 300, 400, 500]

net_present_value(discount_rate1, cash_flow1)

Чистая приведенная стоимость равна: 111.99


In [3]:
discount_rate0 = [15 / 4] * 5

net_present_value(discount_rate0, cash_flow1)

'Размерности списков не равны. Внимательней!'

Из библиотеки ***numpy_financial***

In [4]:
import numpy as np
import numpy_financial as nf

dr_arr1, cf_arr1 = np.array([_ / 100 for _ in discount_rate1]), np.array(cash_flow1)
f'Чистая приведенная стоимость равна : {round(nf.npv(dr_arr1, cf_arr1), 3)}'

'Чистая приведенная стоимость равна : 111.989'

<div class="alert alert-block alert-success"><em>
    <b>Ответ:</b><br>
        Чистая приведенная стоимость равна :<b> 111,99</b>. Положительна. Проект <b>ПРИНИМАЕТСЯ</b>.<br>    
</em></div>

***

<div class="alert alert-block alert-info"><em>
<b>Задача 2.</b> Для инвестиционного проекта с денежными потоками:<br>
Год 0 1 2 3 4 5 6;<br>
денежный поток -1500 100 200 300 400 500 600;<br>
найти NPV, если первые два года ставка дисконтирования равна 20%,<br>
    следующие два года она равна 15%, и затем становится 10%.

Расчёт в таблице ***EXCEL***:

<img src="ДЗ_22_024.png" width="620" height="200">

В ***PYTHON***'е

In [5]:
discount_rate2 = [20, 20, 20, 15, 15, 10, 10]
cash_flow2 = [-1500, 100, 200, 300, 400, 500, 600]

net_present_value(discount_rate2, cash_flow2, 1)

Проект НЕ окупается за данные периоды

Чистая приведенная стоимость равна: -387.52


<div class="alert alert-block alert-success"><em>
    <b>Ответ:</b><br>
        Чистая приведенная стоимость равна: <b>- 387.5</b>.<br>    
</em></div>

***

<div class="alert alert-block alert-info"><em>
<b>Задача 3.</b> Для проекта из задачи 2 найти внутреннюю норму доходности.<br>

По формуле ***EXCEL*** :

<img src="ДЗ_22_031.png" width="620" height="200">

Из библиотеки ***numpy_financial***

In [6]:
f'Внутренняя норма доходности равна: {round(nf.irr(cash_flow2) * 100, 3)} %'

'Внутренняя норма доходности равна: 8.252 %'

<div class="alert alert-block alert-success"><em>
    <b>Ответ:</b><br>
        Внутренняя норма доходности равна:  <b>8.3 %</b>.<br>    
</em></div>

***

<div class="alert alert-block alert-info"><em>
<b>Задача 4.</b>Есть два инвестиционных проекта со следующими денежными потоками:<br>
Год 0 1 2 3 4 5;<br>
Проект А -1000 100 250 450 500 550;<br>
Проект Б -1000 200 300 400 450 500;<br>
Если стоимость денег равна 10%, и инвестор хочет получить максимальную доходность<br>
    на вложенный рубль инвестиций, то какой проект он должен выбрать?

Расчёт в таблице ***EXCEL***:

<img src="ДЗ_22_041.png" width="640" height="200">

В ***PYTHON***'е :

In [7]:
discount_rate4 = [10] * 6
cash_flow4a = [-1000, 100, 250, 450, 500, 550]
cash_flow4b = [-1000, 200, 300, 400, 450, 500]

In [8]:
print('Для проекта А')
net_present_value(discount_rate4, cash_flow4a)

Для проекта А
Чистая приведенная стоимость равна: 318.63


In [9]:
print('Для проекта B')
net_present_value(discount_rate4, cash_flow4b)

Для проекта B
Чистая приведенная стоимость равна: 348.09


<div class="alert alert-block alert-success"><em>
    <b>Ответ:</b><br>
    Чистая приведенная стоимость проекта А меньше, чем у проекта В : <b>318.6 < 348.1</b>.<br>
    А также индекс доходности выше у <b>проекта В</b>.<br>
    <b>Инвестор должен выбрать проект В</b>.<br>    
</em></div>

***

<div class="alert alert-block alert-info"><em>
<b>Задача 5.</b>Есть два инвестиционных проекта со следующими денежными потоками:<br>
Год 0 1 2 3 4 5;<br>
Проект А -1000 100 250 450 500 550;<br>
Проект Б -1000 200 300 400 450 500;<br>
Инвестор хочет выбрать один из проектов по критерию ликвидности с учетом<br> 
    временной стоимости денег в размере 10% годовых. Какой он должен выбрать?<br>

Расчёт в таблице ***EXCEL***:

<img src="ДЗ_22_05.png" width="700" height="200">

В ***PYTHON***'е :

In [10]:
discount_rate5 = [10] * 6
cash_flow5a = [-1000, 100, 250, 450, 500, 550]
cash_flow5b = [-1000, 200, 300, 400, 450, 500]

In [11]:
net_present_value(discount_rate5, cash_flow5a, 2)

Срок окупаемости : 4 года, 0 месяцев, 15 дней;

Чистая приведенная стоимость равна: 318.63


In [12]:
net_present_value(discount_rate5, cash_flow5b, 5)

Срок окупаемости : 3 года, 7 месяцев, 5 дней;

Чистая приведенная стоимость равна: 348.09


<div class="alert alert-block alert-success"><em>
    <b>Ответ:</b><br>
     Срок окупаемости <b>проекта В</b> меньше, выбираем его.<br>    
</em></div>

***